ci/cscs-ci.yml

```diff
@@ -1,185 +1,139 @@
-    - echo "BASE_IMAGE_${PYVERSION_PREFIX}=$PERSIST_IMAGE_NAME" >> build.env
+    # We create a tag that depends on the SHA value of ci/Dockerfile,
+    # the docker build arguments since the Docker file is heavily parameterized,
+    # and the contents of the `uv.lock` with the python dependencies versions.
+    - DOCKER_TAG=`echo "$(cat ${DOCKERFILE}) ${DOCKER_BUILD_ARGS} $(cat uv.lock)" | sha256sum | head -c 16`
+    - export GT4PY_IMAGE_PREFIX=${CSCS_REGISTRY_PATH}/public/${ARCH}/base/gt4py-ci:${DOCKER_TAG}
+    - echo "GT4PY_IMAGE_PREFIX=${GT4PY_IMAGE_PREFIX}" >> base.env
+    - export PERSIST_IMAGE_NAME=${GT4PY_IMAGE_PREFIX}-${PY_VERSION}
+    # Uncomment the following line to see the environment variables for debugging 
+    # - env | sort 
+  parallel:
+    matrix:
+      - PY_VERSION: *test_python_versions
   artifacts:
     reports:
-      dotenv: build.env
-  variables:
-    DOCKERFILE: ci/base.Dockerfile
-    # change to 'always' if you want to rebuild, even if target tag exists already (if-not-exists is the default, i.e. we could also skip the variable)
-    CSCS_REBUILD_POLICY: if-not-exists
-    DOCKER_BUILD_ARGS: '["CUDA_VERSION=$CUDA_VERSION", "CUPY_PACKAGE=$CUPY_PACKAGE", "CUPY_VERSION=$CUPY_VERSION", "UBUNTU_VERSION=$UBUNTU_VERSION", "PYVERSION=$PYVERSION"]'
-.build_baseimage_x86_64:
-  extends: [.container-builder-cscs-zen2, .build_baseimage]
-  variables:
-    CUDA_VERSION: 11.4.3
-    CUPY_PACKAGE: cupy-cuda11x
-    CUPY_VERSION: 12.3.0 # latest version that supports cuda 11
-    UBUNTU_VERSION: 20.04  # 22.04 hangs on daint in some tests for unknown reasons.
-.build_baseimage_aarch64:
-  extends: [.container-builder-cscs-gh200, .build_baseimage]
-  variables:
-    CUDA_VERSION: 12.6.2
-    CUPY_PACKAGE: cupy-cuda12x
-    CUPY_VERSION: 13.4.1
-    UBUNTU_VERSION: 22.04
-
-# build_py313_baseimage_x86_64:
-#   extends: .build_baseimage_x86_64
-#   variables:
-#     <<: *py313
-build_py313_baseimage_aarch64:
-  extends: .build_baseimage_aarch64
-  variables:
-    <<: *py313
-
-# build_py310_baseimage_x86_64:
-#   extends: .build_baseimage_x86_64
-#   variables:
-#     <<: *py310
-build_py310_baseimage_aarch64:
-  extends: .build_baseimage_aarch64
-  variables:
-    <<: *py310
-
+      dotenv: base.env
 
-.build_image:
-  stage: image
+.build_extra_cuda:
   variables:
-    # make sure we use a unique name here, otherwise we could create a race condition, when multiple pipelines
-    # are running.
-    PERSIST_IMAGE_NAME: $CSCS_REGISTRY_PATH/public/$ARCH/gt4py/gt4py-ci:$CI_COMMIT_SHA-$PYVERSION
-    DOCKERFILE: ci/checkout.Dockerfile
-    DOCKER_BUILD_ARGS: '["PYVERSION=$PYVERSION", "BASE_IMAGE=${BASE_IMAGE_${PYVERSION_PREFIX}}"]'
-.build_image_x86_64:
-  extends: [.container-builder-cscs-zen2, .build_image]
-.build_image_aarch64:
-  extends: [.container-builder-cscs-gh200, .build_image]
-
+    BASE_IMAGE: docker.io/nvidia/cuda:${CUDA_VERSION}-devel-ubuntu${UBUNTU_VERSION}
```

Comment:
This could be `jfrog.svc.cscs.ch/dockerhub/nvidia/cuda:${CUDA_VERSION}-devel-ubuntu${UBUNTU_VERSION}`, which is proxied through JFrog, and cached. Network is then all CSCS internal and probably faster.